kernel/src/main/java/org/kframework/parser/concrete2kore/disambiguation/TypeInferenceVisitor.java

```diff
@@ -0,0 +1,244 @@
+// Copyright (c) 2019 K Team. All Rights Reserved.
+package org.kframework.parser.concrete2kore.disambiguation;
+
+import org.kframework.builtin.Sorts;
+import org.kframework.definition.Production;
+import org.kframework.definition.NonTerminal;
+import org.kframework.kore.Sort;
+import org.kframework.parser.Ambiguity;
+import org.kframework.parser.Constant;
+import org.kframework.parser.ProductionReference;
+import org.kframework.parser.Term;
+import org.kframework.parser.TermCons;
+import org.kframework.parser.SetsTransformerWithErrors;
+import org.kframework.utils.errorsystem.KEMException;
+import org.kframework.utils.errorsystem.KException;
+import org.kframework.utils.errorsystem.KException.ExceptionType;
+import org.kframework.utils.errorsystem.KException.KExceptionGroup;
+import org.kframework.utils.errorsystem.ParseFailedException;
+import org.kframework.utils.errorsystem.VariableTypeClashException;
+
+import org.pcollections.ConsPStack;
+
+import scala.util.Either;
+import scala.util.Left;
+import scala.util.Right;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static org.kframework.Collections.*;
+import static org.kframework.kore.KORE.*;
+
+public class TypeInferenceVisitor extends SetsTransformerWithErrors<ParseFailedException> {
```

Comment:
1. This file contains indentation with both 2 spaces and 4. Choose one.
2. There is a lot of new code here. Please add some overview comments and more inline comments.